File: BOJ/9012.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main(void) {
	int T;
	string a;
	
	cin >> T;
	while (T--) {
		cin >> a;
		stack<char> st;
		for (int i = 0; i < a.length(); i++) {
			if (a[i] == '(') st.push(a[i]);
			else if (!st.empty()) st.pop();
			else {
				cout << "NO" << '\n';
				break;
			}
			if (i == a.length() - 1) {
				if (!st.empty()) cout << "NO" << '\n';
				else cout << "YES" << '\n';
			}
		}
	}
	return 0;
}

File: BOJ/9012.py
import sys
n = int(input())

def is_v(ss):
    stack = []
    for s in ss:
        if s=='(':
            stack.append(0)
        else:
            if stack:
                stack.pop()
            else:
                return False
    if stack:
        return False
    return True
        
for _ in range(n):
    s = sys.stdin.readline().rstrip()
    if is_v(s):
        print('YES')
    else:
        print('NO')
